src/common/get_messages_in_datasets.py

```diff
@@ -0,0 +1,110 @@
+
+            log.info(f"Downloaded {len(updated_messages)} updated messages in this dataset, and "
+                     f"{len(ws_corrected_messages)} messages that were previously in this dataset but have moved.")
+
+            # Update the latest seen ws message from this dataset
+            if len(ws_corrected_messages) > 0:
+                for msg in ws_corrected_messages:
+                    if latest_ws_message_timestamp is None or msg.last_updated > latest_ws_message_timestamp:
+                        latest_ws_message_timestamp = msg.last_updated
+                cache.set_date_time(f"{engagement_db_dataset}_ws", latest_ws_message_timestamp)
+
+            cache_messages = cache.get_messages(engagement_db_dataset)
+            for msg in cache_messages:
+                if msg.message_id in {msg.message_id for msg in ws_corrected_messages}:
+                    continue
+                messages.append(msg)
+        else:
+            log.warning(f"Performing a full download for {engagement_db_dataset} messages...")
+
+            full_download_filter = lambda q: q \
+                .where("dataset", "==", engagement_db_dataset)
+
+            messages = engagement_db.get_messages(firestore_query_filter=full_download_filter)
+            log.info(f"Downloaded {len(messages)} messages")
+
+        # Filter messages for their latest versions
+        # (Filter by sorting in reverse order of last_updated and keeping the first snapshot we see of each id)
+        latest_messages = []
+        seen_message_ids = set()
+        messages.sort(key=lambda msg: msg.last_updated, reverse=True)
+        for msg in messages:
+            if msg.message_id not in seen_message_ids:
+                seen_message_ids.add(msg.message_id)
+                latest_messages.append(msg)
+
+        log.info(f"Filtered for latest message snapshots: {len(latest_messages)}/{len(messages)} snapshots remain")
+        messages = latest_messages
+        engagement_db_messages_map[engagement_db_dataset] = messages
+
+        # Update latest_message_timestamp
+        for msg in messages:
+            msg_last_updated = msg.last_updated
+            if latest_message_timestamp is None or msg_last_updated > latest_message_timestamp:
+                latest_message_timestamp = msg_last_updated
+
+        if cache is not None and latest_message_timestamp is not None:
+            # Export latest message timestamp to cache.
+            if latest_message_timestamp is not None:
+                cache.set_date_time(engagement_db_dataset, latest_message_timestamp)
+
+            if full_download_required:
+                # Export this as the ws case too, as there will be no need to check for ws messages that moved from
+                # this dataset before this initial fetch.
+                cache.set_date_time(f"{engagement_db_dataset}_ws", latest_message_timestamp)
+
+            # Export project engagement_dataset files
+            if len(messages) > 0:
```

Comment:
I think we still want to catch duplicate origin_ids here.

Reply:
Thanks for catching this, I think the code on main changed while I was working on this PR.
I have added this back in, and manually compared both versions to check for any other differences.